File: docs/algorithms/code/algorithms/isValidBST.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBSTMain(
  root: TreeNode | null,
  lower: number,
  upper: number
): boolean {
  if (!root) {
    return true;
  }

  if (root.val <= lower || root.val >= upper) {
    return false;
  }

  return (
    isValidBSTMain(root.left, lower, root.val) &&
    isValidBSTMain(root.right, root.val, upper)
  );
}

function isValidBST(root: TreeNode | null): boolean {
  return isValidBSTMain(root, -Infinity, Infinity);
}

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
  const stack: TreeNode[] = [];
  let prevValue = -Infinity;

  while (stack.length || root) {
    while (root) {
      stack.push(root);
      root = root.left;
    }

    root = stack.pop();

    if (root.val <= prevValue) {
      return false;
    }

    prevValue = root.val;

    root = root.right;
  }

  return true;
}

File: docs/algorithms/code/algorithms/isValidBST.go
// 思路：通过递归判断每个节点是否符合二叉搜索树的定义
// 复杂度分析：时间复杂度O(n)，空间复杂度O(h)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isValidBSTHelper(root *TreeNode, left int, right int) bool {
	if root == nil {
		return true
	}

	if root.Val <= left || root.Val >= right {
		return false
	}

	return isValidBSTHelper(root.Left, left, root.Val) && isValidBSTHelper(root.Right, root.Val, right)
}

func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return isValidBSTHelper(root.Left, -1<<31-1, root.Val) && isValidBSTHelper(root.Right, root.Val, 1<<31)
}